testing/integration_testing/provisioning.py

```diff
@@ -0,0 +1,141 @@
+# limitations under the License.
+
+"""Helper module for dealing with mobile provisioning profiles.
+
+This module provides functionality to extract a UUID from a mobile provisioning
+file and patch it into an exports plist. The motivation for this arose when it
+was discovered that the UUID changes each time the profile is updated. Since the
+profile was pulled from a source where it was updated daily, manually specifying
+the UUID in the exports plist was not feasible.
+
+The UUID is needed to tell xcodebuild which provisioning profile to use.
+
+This should only be run on MacOS, as it uses a mac-specific
+application.
+
+"""
+
+import plistlib
+import subprocess
+
+
+def patch_provisioning_profile(plist_path, bundle_id, uuid):
+  """Patches the plist with a provisioning profile.
+
+  Will add an entry to the provisioning profiles dictionary in the provided
+  plist. The entry will be bundle_id:uuid, associating a project with a
+  provisioning profile. This is needed for exporting an xcode archive using
+  xcodebuild.
+
+  If an entry with the same bundle id already exists, its value will be
+  overwritten.
+
+  Args:
+    plist_path (str): Path to a valid plist with a provisioningProfiles
+        dictionary in its root.
+    bundle_id (str): Bundle id identifying this project.
+    uuid (str): UUID from a mobile provisioning profile.
+
+  Raises:
+    ValueError: plist not in correct format.
+
+  """
+  # Need to explicitly specify binary format for load to work.
+  with open(plist_path, "rb") as f_read:
+    plist = plistlib.load(f_read)
+  _patch_provisioning_profile(plist, bundle_id, uuid)
+  with open(plist_path, "wb") as f_write:
+    plistlib.dump(plist, f_write)
+
+
+def _patch_provisioning_profile(plist, bundle_id, uuid):
+  """Patches the plist dictionary with bundle_id:uuid."""
+  try:
+    profiles = plist["provisioningProfiles"]
+    profiles[bundle_id] = uuid
+  except (TypeError, KeyError):
+    raise ValueError(
+        "Plist not in correct format. 'provisioningProfiles' misplaced?")
+
+
+def get_provision_id(provision_path):
+  """Extracts the provisioning UUID in the provided provisioning profile.
+
+  Can only be run on macOS.
+
+  Args:
+    provision_path (str): Path to a .mobileprovision file.
+
+  Returns:
+    A string corresponding to a UUID.
+
+  Raises:
+    ValueError: If provision_path is empty or None, or isn't a mobileprovision.
+    RuntimeError: Called this script on anything but a Mac.
+
+  """
+  if not provision_path:
+    raise ValueError("Must provide valid directory for provisions")
+
+  if not provision_path.endswith(".mobileprovision"):
+    raise ValueError("Path is not a .mobileprovision file: %s" % provision_path)
+
+  plist = _extract_plist(provision_path)
+  uuid = _extract_uuid(plist)
+  if not uuid:
+    raise RuntimeError("No UUID found in provision %s." % provision_path)
+  return uuid
+
+
+def _extract_plist(provision_path):
```

Comment:
Just a note that a lot of this mobileprovision-related stuff won't be usable on GHA as-is. It might be worth stripping all these out until we have a solution.

Reply:
Right, I was somewhat ambivalent about whether to include the iOS logic (which is half the code in this PR). My main reason for including it is to provide a starting/reference point for when this is modified to handle whatever solution we come up with for iOS, especially since that person might not be me (since Mou is looking into signing, he might also wire things up for this), and also since there are quite a few changes here relative to what's in google3.

Reply:
Yeah I think that makes sense. Let the record show that it doesn't work as-is now. :)